back/app/models/user.rb

```diff
@@ -70,10 +70,23 @@ def enabled_roles
       ['admin']
     end
 
+    # Returns the user record from the database which matches the specified
+    # email address (case-insensitive) or `nil`.
+    # @param email [String] The email of the user
+    # @return [User, nil] The user record or `nil` if none could be found.
     def find_by_cimail(email)
       where('lower(email) = lower(?)', email).first
     end
 
+
+    # Returns the user record from the database which matches the specified
+    # email address (case-insensitive) or raises `ActiveRecord::RecordNotFound`.
+    # @param email [String] The email of the user
+    # @return [User] The user record
+    def find_by_cimail!(email)
```

Comment:
Nice method! It's intuitive and Rails-way :+1: 

But maybe I wouldn't add comments for it.

1. The method is really short. It's much faster to read the code than the comments.
2. It becomes not DRY, not that Ruby-way.
    1. The comments duplicate the code. `raises ActiveRecord::RecordNotFound`, `@param email` - it's already in the code. 
    2. It's hard to imagine that email is not a string. Or that `user` is not `User` from `generate_reset_password_token`. 
3. In the future the comments that duplicate the code are likely to become outdated. So, they will only confuse.

If you use IDE, indeed it may help to write code. But I would at least remove the description and leave only return and param types.

But maybe it's just my personal opinion, so you can ignore it :slightly_smiling_face:

Reply:
I agree that this might be a bit overkill in terms of documention, but I prefer to have a bit too much than too little I think. 

I indeed use [solargraph](https://github.com/castwide/solargraph) as a language LSP server for Ruby to automatically show completions / documentation for me and I find it super helpful to have it available because I do not have to look at the implementation anymore! :)

![image](https://user-images.githubusercontent.com/866318/143553231-1eed9a4c-6ddc-4721-976c-91be857ef778.png)

Reply:
I see. Makes sense then :+1:

Reply:
Solargraph looks really nice, thanks for the tip!

Do you configure it somehow? Or it's just gem + editor extension?

My `Solargraph Environment Info` says `Using Bundler: false`, but somehow it knows about ActiveRecord. Should I tell Solargraph about my Gemfile?

Reply:
I've done the following:

in `citizenlab/back`: 

```shell
# Download & install solargraph
$ gem install solargraph

# Download Ruby core documentation
$ solargraph download-core

# Generate documentation for bundled gems
$ solargraph bundle 
```

The setup is a bit wonky because the Ruby version & citizenlab application  is running in a docker container, but my editor (Emacs) needs solargraph on my host machine to pick it up.
So what I did is simply install and use the same Ruby version that we have in our docker container (2.5.3) and bundle the dependencies also locally.

From that point on my editor was able to show completions etc.

Reply:
Thanks for the extensive answer! Maybe we could add solar to our git repo together with this Rails file https://solargraph.org/guides/rails

Added this point to CDM doc, so we don't forget to discuss it later.